### PR TITLE
Issue#156 Manage contacts on cooperations

### LINKED ISSUE #156: Manage contacts on /cooperations
**As a** cooperation administrator **I want to** have the ability to manage contacts for the cooperation so I can edit, create or delete them.

### Description
After implementing https://github.com/ita-social-projects/Home/issues/91 we have the ability to provide contacts for cooperation when a system administrator creates cooperation. But as a cooperation administrator, I want to be able to manage contacts on my own. Please provide API for it.

### Acceptance Criteria
I have enough API for managing contacts for cooperation.

### Epic link
Epic #3 
[ZenHub Epic](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/55)

### Tasks 
- [ ] TBD
- [ ] ...

### PR DESCRIPTION
**Pull request by likeRewca**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/156)

## Code reviewers

- [ ] @kh0ma
- [ ] @tex1988
- [ ] @Ilya-Ross
- [ ] @VadymSokorenko
- [ ] @denis-litvinov
- [ ] @annaprok

### Second Level Review

- [ ] @github_username

## Summary of issue

Provide the ability to create and manage contacts to the admin.

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue156.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    </changeSet>
+
+    <changeSet id="Issue156.004" author="likeRewca-Yura">
+        <addUniqueConstraint schemaName="public"
+                             tableName="users"
+                             constraintName="user_email_key"
+                             columnNames="email"/>
+    </changeSet>
+
+</databaseChangeLog>
```
Comment (kh0ma) [inline home-data-migration/src/main/resources/db/changelog/0.0.1/Issue156.xml]: Please do not forget to add newline at the end of the file.

File: home-open-api/src/main/resources/yaml/openapi.yaml [hunk 2]
```diff
@@ -1640,4 +1902,4 @@ components:
           type: integer
           example: 12
         address:
-          $ref: '#/components/schemas/Address'
+          $ref: '#/components/schemas/Address'
```
Comment (kh0ma) [inline home-open-api/src/main/resources/yaml/openapi.yaml]: Add newline at the end of the file.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/BaseContactService.java [hunk 3]
```diff
@@ -25,50 +21,30 @@
...
-    }
+    protected abstract void checkAndFillParentEntity(ContactDto contactDto, Contact createContact, Long parentEntityId);
 
     @Override
-    public ContactDto updateContact(Long id, ContactDto updateContactDto) {
-        Contact contact = contactRepository.findById(id)
-            .filter(Contact::getEnabled)
-            .orElseThrow(() -> new NotFoundHomeException("User with id:" + id + " is not found"));
+    public ContactDto updateContact(Long parentEntityId, Long contactId, ContactDto updateContactDto) {
+        var contact = findingAndCheckingContactParentEntity(contactId, parentEntityId);
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/BaseContactService.java]: checkAndGetParentEntity

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                .ifPresent(contact -> {
+                    throw new AlreadyExistHomeException("Cooperation with id "
+                        + parentEntityId + " already has main " + contactDto.getType() + " contact");
+                });
+        }
+
+        createContact.setCooperation(cooperation);
+    }
+
+    private Cooperation getCooperationById(Long cooperationId) {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java]: We have to use a service with the corresponding domain. Please inject CoooperationService and retrieve the Cooperation from there. Also, cooperation has to be checked not only for existence but via `enabled` param.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java [hunk 5]
```diff
@@ -86,6 +96,10 @@ public CooperationDto getCooperationById(Long id) {
             .orElseThrow(() -> new NotFoundHomeException(String.format(NOT_FOUND_COOPERATION_FORMAT, id)));
         List<House> houseList = houseRepository.findHousesByCooperationId(toGet.getId());
         toGet.setHouses(houseList);
+        List<Contact> contactList = new ArrayList<>();
+        contactList.addAll(contactRepository.findAllByCooperationIdAndType(toGet.getId(), ContactType.EMAIL));
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java]: Why do we need to load contacts separately by type? Can we load all of them together?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserContactServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                .findAny()
+                .ifPresent(contact -> {
+                    throw new AlreadyExistHomeException("User with id "
+                        + parentEntityId + " already has main " + contactDto.getType() + " contact");
+                });
+        }
+        createContact.setUser(user);
+    }
+
+    private User getUserById(Long id) {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserContactServiceImpl.java]: The same as for Cooperation. User UserService for it.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -38,9 +39,7 @@
     @Override
     @Transactional
     public UserDto createUser(UserDto createUserDto) {
-        if (userRepository.findByEmail(createUserDto.getEmail()).isPresent()) {
-            throw new AlreadyExistHomeException("User with email" + createUserDto.getEmail() + " is already exists");
-        } else {
+        if (Boolean.FALSE.equals(userRepository.findByEmail(createUserDto.getEmail()).isPresent())) {
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: 🥺
image

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-data-migration/src/main/resources/db/changelog/0.0.1/Issue156.xml (hunk 1 above)
Comment (likeRewca) [reply home-data-migration/src/main/resources/db/changelog/0.0.1/Issue156.xml]: yeap, I forgot about this

File: home-open-api/src/main/resources/yaml/openapi.yaml (hunk 2 above)
Comment (likeRewca) [reply home-open-api/src/main/resources/yaml/openapi.yaml]: saw and corrected

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/BaseContactService.java (hunk 3 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/BaseContactService.java]: I think checkAndGetContactByParentId more better becouse we return Contact.

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java (hunk 5 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationServiceImpl.java]: At that moment I used functionality what I have

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java (hunk 4 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java]: You mean that i need use similar name method from side CooperationServiseImpl?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: it was double check))

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserContactServiceImpl.java (hunk 6 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserContactServiceImpl.java]: done

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -87,6 +84,17 @@ public UserDto updateUser(Long id, UserDto updateUserDto) {
         return mapper.convert(fromDB, UserDto.class);
     }
 
+    private User checkingUniqueEmailParameterForUpdate(Long id, UserDto userDto) {
+        var userFindById = userRepository.findById(id).filter(User::getEnabled)
```
Comment (tex1988) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: I think you have to revork this method in way of email uniqueness validation without returning a user

**Review by tex1988 — CHANGES_REQUESTED** [review]

File: home-open-api/src/main/resources/yaml/openapi.yaml [hunk 9]
```diff
@@ -1107,6 +1107,252 @@ paths:
           $ref: '#/components/responses/NotFound'
         default:
           $ref: '#/components/responses/InternalServerError'
+  /cooperation/{cooperation_id}/contacts:
+    post:
+      tags:
+        - cooperation contact
+      summary: Create new Contact for the Cooperation
+      description: The endpoint creates Contacts for the specified Cooperation ID
+      operationId: createContactsOnCooperation
```
Comment (kh0ma) [inline home-open-api/src/main/resources/yaml/openapi.yaml]: `createContactOnCooperation`

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                        + parentEntityId + " already has main " + contactDto.getType() + " contact");
+                });
+        }
+
+        createContact.setCooperation(cooperation);
+    }
+
+    @Override
+    protected Contact checkAndGetContactByParentId(Long contactId, Long parentEntityId) {
+        return contactRepository.findByIdAndCooperationId(contactId, parentEntityId)
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java]: Shouldn't we check that cooperation is enabled?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    public CooperationContactServiceImpl(ContactRepository contactRepository,
+                                         ServiceMapper mapper,
+                                         CooperationServiceImpl cooperationService) {
+        super(contactRepository, mapper);
+        this.cooperationService = cooperationService;
+    }
+
+    @Override
+    protected void checkAndFillParentEntity(ContactDto contactDto, Contact createContact, Long parentEntityId) {
+        var cooperation = mapper
```
Comment (kh0ma) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java:36]: Shouldn't we check that cooperation is enabled?

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java (hunk 11 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java:36]: in method getCooperationById (in Cooperation service) we check for enabled or not

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java (hunk 10 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/CooperationContactServiceImpl.java]: if we have cooperative and it is not available what kind of exception to issue?
"The cooperation " + cooperation.getName() + " not enabled."?

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java (hunk 8 above)
Comment (likeRewca) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: This method was chosen to reduce the number of duplicate lines. I have to check check multiple event options. And pulling it out of the user base when checking and updating is superfluous. 
If you have any suggestions on how to do this, I will take note of this.

File: home-open-api/src/main/resources/yaml/openapi.yaml (hunk 9 above)
Comment (likeRewca) [reply home-open-api/src/main/resources/yaml/openapi.yaml]: done

**Review by tex1988 — APPROVED** [review]

**Review by kh0ma — APPROVED** [review]

**Review by annaprok — APPROVED** [review]

**Review by IRoss-git — APPROVED** [review]

**Pull request merged**